GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,466 @@
+            else if (string.IsNullOrEmpty(this.nugetUpgraderConfig.FeedUrlForCredentials))
+            {
+                message = "URL to lookup credentials has not been configured";
+                return false;
+            }
+            else
+            {
+                message = null;
+            }
+
+            message = null;
+            return true;
+        }
+
+        public bool TryQueryNewestVersion(out Version newVersion, out string message)
+        {
+            try
+            {
+                IList<IPackageSearchMetadata> queryResults = this.nuGetFeed.QueryFeedAsync(this.nugetUpgraderConfig.PackageFeedName).GetAwaiter().GetResult();
+
+                // Find the latest package
+                IPackageSearchMetadata highestVersion = null;
+                foreach (IPackageSearchMetadata result in queryResults)
+                {
+                    if (highestVersion == null || result.Identity.Version > highestVersion.Identity.Version)
+                    {
+                        highestVersion = result;
+                    }
+                }
+
+                if (highestVersion != null &&
+                    highestVersion.Identity.Version.Version > this.installedVersion)
+                {
+                    this.latestVersion = highestVersion;
+                }
+
+                newVersion = this.latestVersion?.Identity?.Version?.Version;
+
+                if (newVersion != null)
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - new version available: installedVersion: {this.installedVersion}, latestAvailableVersion: {highestVersion}");
+                    message = $"New version {highestVersion.Identity.Version} is available.";
+                    return true;
+                }
+                else if (highestVersion != null)
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - up-to-date");
+                    message = $"Latest available version is {highestVersion.Identity.Version}, you are up-to-date";
+                    return true;
+                }
+                else
+                {
+                    this.tracer.RelatedInfo($"{nameof(this.TryQueryNewestVersion)} - no versions available from feed.");
+                    message = $"No versions available via feed.";
+                }
+            }
+            catch (Exception ex)
+            {
+                this.tracer.RelatedError($"{nameof(this.TryQueryNewestVersion)} failed with: {ex.Message}");
+                message = ex.Message;
+                newVersion = null;
+            }
+
+            return false;
+        }
+
+        public bool TryDownloadNewestVersion(out string errorMessage)
+        {
+            if (this.latestVersion == null)
+            {
+                errorMessage = "No new version to download. Query for latest version to ensure a new version is available before downloading.";

```

Comment:
Is this a user facing error message?  If so, what action would\should a user take if they see this message?

Reply:
I think if you want to stick with the error message approach it could be simplified to: ""No new version to download" with a comment that provides the additional details.

If this will *only* fail due to programmer error than something like an InvalidOperationException is an option too.

Reply:
@mjcheetham I think we agree on what the high level behavior is for this scenario (i.e. logging and stopping the operation). I don't know if introducing a second error handling pattern at this point is the way to achieve that, given our current patterns. An assert would achieve much of the same effect (crash in debug builds, handled more gracefully in product builds).

I am open to opinions on this though...

Reply:
If a failure here results in an obvious failure for the user I'm not sure that the assert would add much value.  I think we could just stick with the error message.  (Either way if a dev sees it they'd investigate)